src/pq.cpp

```diff
@@ -776,24 +828,32 @@ int generate_opq_pivots(const float *passed_train_data, size_t num_train, uint32
 // Output is stored in the in-memory buffer pq.
 // Simplification is based on the following assumptions:
 //   supporting only float data type
-//   dim % num_pq_chunks == 0, which results in a fixed chunk_size
 //   num_centers == 256 by default
-//   make_zero_mean is false by default.
 // These assumptions allow to make the function much simpler and avoid using
 // array of chunk_offsets and centroids.
 // The compiler pragma for multi-threading support is removed from this implementation
 // for the purpose of integration into systems that strictly control resource allocation.
-int generate_pq_data_from_pivots_simplified(const float *data, const size_t num, const float *pivot_data,
+int generate_pq_data_from_pivots_simplified(float *data, const size_t num, const float *pivot_data,
                                             const size_t pivots_num, const size_t dim, const size_t num_pq_chunks,
-                                            std::vector<uint8_t> &pq)
+                                            const std::vector<float> &centroids, const std::vector<uint32_t> &offsets,
+                                            std::vector<PQ_DATA_TYPE> &pq)
 {
-    if (num_pq_chunks == 0 || num_pq_chunks > dim || dim % num_pq_chunks != 0)
+    if (num_pq_chunks < 2 || num_pq_chunks > dim)
```

Comment:
why if < 2 we return -1?

Reply:
Return value -1 means functions failed for some reason.

We check input parameters:
 - num_pq_chunks == 0   =>  there is a division by num_pq_chunks a bit later, which will crash the system
 - num_pq_chunks == 1   =>  originally I had offsets.size() == num_pq_chunks, and then I use offsets[1] as the largest possible size for a chunk. If num_pq_chunks == 1, then access offsets[1] will crash the system. Eventually I accepted the wisdom of the original function and added one more element to offsets so offsets.size() == num_pq_chunks+1 but the check of num_pq_chunks value was left the same.

In general, num_pq_chunks==1 does not make sense anyway, right? In this case the whole vector is squashed into a single number 0..255. Too much information is lost. I can't imagine a practical case when num_pq_chunks==1 is useful.

On the other hand, there are many things that I cannot imagine but they are still valid. :)
I can change the check to if (num_pq_chunks == 0) return -1; if required.

Reply:
I'd say lets do what the last option is, we can allow num_pq_chunks =1.